src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/ordinal/OrdinalRewriterRule.java

```diff
@@ -0,0 +1,210 @@
+        return true;
+    }
+
+    @Override
+    public void rewrite(SQLQueryExpr root) {
+        changeOrdinalAliasInGroupAndOrderBy(root);
+    }
+
+    private boolean hasGroupByWithOrdinals(MySqlSelectQueryBlock query) {
+        if (query.getGroupBy() == null) {
+            return false;
+        } else if (query.getGroupBy().getItems().isEmpty()){
+            return false;
+        }
+
+        return query.getGroupBy().getItems().stream().anyMatch(x ->
+            x instanceof MySqlSelectGroupByExpr && ((MySqlSelectGroupByExpr) x).getExpr() instanceof SQLIntegerExpr
+        );
+    }
+
+    private boolean hasOrderByWithOrdinals(MySqlSelectQueryBlock query) {
+        if (query.getOrderBy() == null) {
+            return false;
+        } else if (query.getOrderBy().getItems().isEmpty()){
+            return false;
+        }
+
+        return query.getOrderBy().getItems().stream().anyMatch(x ->
+            x.getExpr() instanceof SQLIntegerExpr
+            || (
+                x.getExpr() instanceof SQLBinaryOpExpr
+                && ((SQLBinaryOpExpr) x.getExpr()).getLeft() instanceof SQLIntegerExpr
+            )
+        );
+    }
+
+    private void collectOrdinalAliases(SQLQueryExpr root, Map<Integer, SQLExpr> map) {
+        root.accept(new MySqlASTVisitorAdapter() {
+            private Integer count = 0;
+
+            @Override
+            public boolean visit(SQLSelectItem select) {
+                map.put(++count, select.getExpr());
+                return false;
+            }
+        });
+    }
+
+    private void changeOrdinalAliasInGroupAndOrderBy(SQLQueryExpr root) {
+        root.accept(new MySqlASTVisitorAdapter() {
+
+            private String groupException = "Invalid ordinal [%s] specified in [GROUP BY %s]";
+            private String orderException = "Invalid ordinal [%s] specified in [ORDER BY %s]";
+
+            @Override
+            public boolean visit(MySqlSelectGroupByExpr groupByExpr) {
+                SQLExpr expr = groupByExpr.getExpr();
+                if (expr instanceof SQLIntegerExpr) {
+                    Integer ordinalValue = ((SQLIntegerExpr) expr).getNumber().intValue();
+                    SQLExpr newExpr = groupOrdinalToExpr.get(ordinalValue);
+                    checkInvalidOrdinal(newExpr, ordinalValue, groupException);
+                    groupByExpr.setExpr(newExpr);
+                    newExpr.setParent(groupByExpr);
+                }
+                return false;
+            }
+
+            @Override
+            public boolean visit(SQLSelectOrderByItem orderByItem) {
+                SQLExpr expr = orderByItem.getExpr();
+                Integer ordinalValue;
+
+                if (expr instanceof SQLIntegerExpr) {
+                    ordinalValue = ((SQLIntegerExpr) expr).getNumber().intValue();
+                    SQLExpr newExpr = orderOrdinalToExpr.get(ordinalValue);
+                    checkInvalidOrdinal(newExpr, ordinalValue, orderException);
```

Comment:
this is a bit repetitive, you can extract this to method like

    getExpressionByOrdinal

and throw an exception inside (will simplify code a bt)

Reply:
Refactored it into `checkAndGet` method.